File: ollama/codigo-final/pidigits/codegemma-7b/cot/temp_0.7/pidigits.py
import sys

def next_step(q, r, t):
    k = k + 1
    r = q * 4 + 2 * r
    t = 2 * t * k + 1
    q = r // t
    return q, r, t

def extract(x):
    return (q * x + r) // t

def produce(d):
    global N
    if d == extract(3):
        print(d, end=" ")
        N += 1
        if N % 10 == 0:
            print(f"\t:{N}")
    elif d == extract(4):
        while True:
            d = extract(3)
            if d == extract(4):
                print(d, end=" ")
                N += 1
                if N % 10 == 0:
                    print(f"\t:{N}")
                break

# Input validation
try:
    N = int(sys.argv[1])
except ValueError:
    print("Error: Invalid input. Please provide a positive integer.")
    sys.exit(1)

if N <= 0:
    print("Error: Invalid input. Please provide a positive integer.")
    sys.exit(1)

# Initialization
q, r, t = 1, 0, 1
k = 0

# Algorithm loop
while q > r:
    q, r, t = next_step(q, r, t)
    produce(extract(3))
    produce(extract(4))

# Output formatting
print()  # Remove trailing newline if necessary

File: ollama/codigo-final/pidigits/stable-code-3b/cot/temp_0/pidigits.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <gmpxx.h> // Include GMP library

using namespace std;

int main(int argc, char *argv[]) {
    // Error handling for command-line arguments
    if (argc != 2) {
        cerr << "Usage: pidigits N" << endl;
        return 1;
    }

    int n = atoi(argv[1])); // Convert command-line argument to integer

    if (n <= 0) {
        cerr << "N must be a positive integer." << endl;
        return 1;
    }

    // Initial state variables
    mpz_class q = 1, r = 0, t = 1, k = 0;
    int digits_calculated = 0;

    // Main loop to calculate digits of Pi
    while (q <= r) {
        k++;
        r += 2 * q;
        r *= (2 * k + 1));
        t *= (2 * k + 1));
        q *= k;
        digits_calculated++;

        // Extract digits and verify
        mpz_class d3 = extract(3);
        mpz_class d4 = extract(4);

        if (d3 == d4) {
            cout << d3 << "\t: " << n << endl;
            produce(d3);
        }
    }

    return 0;
}

// Helper functions for the Gibbons spigot algorithm
mpz_class extract(int index) {
    mpz_class result;
    if (index == 3) {
        result = (q * 3 + r) / t;
    } else if (index == 4) {
        result = (q * 4 + r) / t;
    } else {
        cerr << "Invalid index" << endl;
        return 0; // Handle invalid input
    }
    return result;
}

void produce(mpz_class digit) {
    q *= 10;
    r = (r - t * digit) * 10;
}
